boa/src/builtins/json/mod.rs

```diff
@@ -72,59 +72,136 @@ impl Json {
+        let unfiltered = context.eval(script_string.as_bytes())?;
+
+        match args.get(1).cloned().unwrap_or_default().as_object() {
+            // 11. If IsCallable(reviver) is true, then
+            Some(obj) if obj.is_callable() => {
+                // a. Let root be ! OrdinaryObjectCreate(%Object.prototype%).
+                let root = context.construct_object();
+
+                // b. Let rootName be the empty String.
+                // c. Perform ! CreateDataPropertyOrThrow(root, rootName, unfiltered).
+                root.create_data_property_or_throw("", unfiltered, context)
+                    .expect("CreateDataPropertyOrThrow should never throw here");
+
+                // d. Return ? InternalizeJSONProperty(root, rootName, reviver).
+                Self::internalize_json_property(root, "".into(), obj, context)
             }
-            Err(err) => context.throw_syntax_error(err.to_string()),
+            // 12. Else,
+            // a. Return unfiltered.
+            _ => Ok(unfiltered),
         }
     }
 
-    /// This is a translation of the [Polyfill implementation][polyfill]
+    /// `25.5.1.1 InternalizeJSONProperty ( holder, name, reviver )`
     ///
-    /// This function recursively walks the structure, passing each key-value pair to the reviver function
-    /// for possible transformation.
+    /// More information:
+    ///  - [ECMAScript reference][spec]
     ///
-    /// [polyfill]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse
-    fn walk(
-        reviver: &JsValue,
+    /// [spec]: https://tc39.es/ecma262/#sec-internalizejsonproperty
+    fn internalize_json_property(
+        holder: JsObject,
+        name: JsString,
+        reviver: JsObject,
         context: &mut Context,
-        holder: &mut JsValue,
-        key: &PropertyKey,
     ) -> JsResult<JsValue> {
-        let value = holder.get_field(key.clone(), context)?;
-
-        if let JsValue::Object(ref object) = value {
-            let keys: Vec<_> = object.borrow().properties().keys().collect();
+        // 1. Let val be ? Get(holder, name).
+        let val = holder.get(name.clone(), context)?;
+
+        // 2. If Type(val) is Object, then
+        if let Some(obj) = val.as_object() {
+            // a. Let isArray be ? IsArray(val).
+            // b. If isArray is true, then
+            if obj.is_array() {
+                // i. Let I be 0.
+                // ii. Let len be ? LengthOfArrayLike(val).
+                let mut i = 0;
+                let len = obj.length_of_array_like(context)? as i64;
+
+                // iii. Repeat, while I < len,
+                while i < len {
+                    // 1. Let prop be ! ToString(𝔽(I)).
+                    // 2. Let newElement be ? InternalizeJSONProperty(val, prop, reviver).
+                    let new_element = Self::internalize_json_property(
+                        obj.clone(),
+                        i.to_string().into(),
```

Comment:
Hmm... I'm doubtful if applying `to_string` to an indexed key is a good idea. Can you confirm if making `name` a `Into<PropertyKey>` instead of a `JsString` solves any tests?

Reply:
`name` must be some kind of string for the `call` at the end of the function. If I change it into a `Into<PropertyKey>`, the `JsValue` the gets passed to the `call` function would be numeric in some cases, because `PropertyKey.into::<JsValue>()` may convert into a numeric `JsValue`.

Reply:
The problem I see is that when you call `internalize_json_property` recursively for an array, it will try to `Get` the property index i, but converted into a string, so it will try to get the string property "i". However, for an array we store indices as an index property, so the `Get` would return `JsValue::Undefined`, I think.

Reply:
Ah I see what you mean. But we do not have to worry about this here. The `From<JsString> for PropertyKey` implementation trys to parse the string into an u32 and returning it as `PropertyKey::Index(u32)`. If that fails it returns is as an `PropertyKey::String(JsString)`.

Reply:
I see. That's okay then. I'd like to avoid casting an index to a string and then back into an index but maybe that'll make the code less readable.